app/auth/oauth_client.py

```diff
@@ -0,0 +1,129 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import sys
+import time
+
+from oauthlib.oauth2 import WebApplicationClient
+from oauthlib.oauth2.rfc6749.errors import OAuth2Error
+from requests_oauthlib import OAuth2Session
+
+from .oauth_provider_app import OAuthProviderApp, JupyterHubProviderApp
+
+
+class RenkuWebApplicationClient(WebApplicationClient):
+    """WebApplicationClientClass enriched with provider/app information and
+    methods for (de-)serializing and obtaining tokens from the provider."""
+
+    def __init__(
+        self,
+        *args,
+        provider_app=None,
+        scopes=[],
```

Comment:
Default argument value should not be mutable - this could lead to various problems. Maybe we should consider it to make it `None` or a `tuple`

Reply:
`None` is the sensible default